modules/renter/streamcache.go

```diff
@@ -0,0 +1,181 @@
+package renter
+
+import (
+	"container/heap"
+	"sync"
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/errors"
+)
+
+// streamHeap is a priority queue and implements heap.Interface and holds chunkData
+type streamHeap []*chunkData
+
+// chunkData contatins the data and the timestamp for the unfinished
+// download chunks
+type chunkData struct {
+	id         string
+	data       []byte
+	lastAccess time.Time
+	index      int
+}
+
+// streamCache contains a streamMap for quick look up and a streamHeap for
+// quick removal of old chunks
+type streamCache struct {
+	streamMap  map[string]*chunkData
+	streamHeap streamHeap
+	cacheSize  uint64
+	mu         sync.Mutex
+}
+
+// Required functions for use of heap for streamHeap
+func (sh streamHeap) Len() int { return len(sh) }
+
+// Less returns the lessor of two elements
+func (sh streamHeap) Less(i, j int) bool { return sh[i].lastAccess.Before(sh[j].lastAccess) }
+
+// Swap swaps two elements from the heap
+func (sh streamHeap) Swap(i, j int) {
+	sh[i], sh[j] = sh[j], sh[i]
+	sh[i].index = i
+	sh[j].index = j
+}
+
+// Push adds an element to the heap
+func (sh *streamHeap) Push(x interface{}) {
+	n := len(*sh)
+	chunkData := x.(*chunkData)
+	chunkData.index = n
+	*sh = append(*sh, chunkData)
+}
+
+// Pop removes element from the heap
+func (sh *streamHeap) Pop() interface{} {
+	old := *sh
+	n := len(old)
+	chunkData := old[n-1]
+	chunkData.index = -1 // for safety
+	*sh = old[0 : n-1]
+	return chunkData
+}
+
+// update updates the heap and reorders
+func (sh *streamHeap) update(cd *chunkData, id string, data []byte, lastAccess time.Time) {
+	cd.id = id
+	cd.data = data
+	cd.lastAccess = lastAccess
+	heap.Fix(sh, cd.index)
+}
+
+// Add adds the chunk to the cache if the download is a streaming
+// endpoint download.
+// TODO this won't be necessary anymore once we have partial downloads.
+func (sc *streamCache) Add(cacheID string, data []byte) {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+
+	// pruning cache to cacheSize - 1 to make room to add the new chunk
+	sc.pruneCache(sc.cacheSize - 1)
+
+	// Add chunk to Map and Heap
+	cd := &chunkData{
+		id:         cacheID,
+		data:       data,
+		lastAccess: time.Now(),
+	}
+	sc.streamMap[cacheID] = cd
+	heap.Push(&sc.streamHeap, cd)
+	sc.streamHeap.update(cd, cd.id, cd.data, cd.lastAccess)
+}
+
+// init initializes the streamCache
+func (sc *streamCache) Init() {
+	sc.mu.Lock()
+	defer sc.mu.Unlock()
+	if sc.cacheSize > 0 {
+		build.Critical("streamCache already initialized")
+		return
+	}
+
+	sc.streamMap = make(map[string]*chunkData)
+	sc.streamHeap = make(streamHeap, 0, defaultStreamCacheSize)
+	sc.cacheSize = defaultStreamCacheSize
+	heap.Init(&sc.streamHeap)
+}
+
+// pruneCache prunes the cache until it is the length of cacheSize
+func (sc *streamCache) pruneCache(size uint64) {
+	for len(sc.streamMap) > int(size) {
```

Comment:
Just to be safe we should add a sanity check that makes sure that `size` is not smaller than `0`. Otherwise we prune forever.

Reply:
Nevermind. It's a `uint` so it should be fine.